src/Microsoft.NetCore.Analyzers/Core/ImmutableCollections/DoNotCallToImmutableCollectionOnAnImmutableCollectionValue.cs

```diff
@@ -55,11 +55,31 @@ public override void Initialize(AnalysisContext context)
             context.RegisterCompilationStartAction(compilationStartContext =>
             {
                 var compilation = compilationStartContext.Compilation;
-                if (compilation.GetTypeByMetadataName(ImmutableArrayMetadataName) == null)
+                var immutableArraySymbol = compilation.GetTypeByMetadataName(ImmutableArrayMetadataName);
+                if (immutableArraySymbol is null)
+                {
+                    var systemNamespace = compilation.GlobalNamespace.GetMembers(nameof(System)).OfType<INamespaceSymbol>().SingleOrDefault();
+                    var systemCollectionsNamespace = systemNamespace?.GetMembers(nameof(System.Collections)).OfType<INamespaceSymbol>().SingleOrDefault();
+                    var systemCollectionsImmutableNamespace = systemCollectionsNamespace?.GetMembers(nameof(System.Collections.Immutable)).OfType<INamespaceSymbol>().SingleOrDefault();
+                    if (systemCollectionsImmutableNamespace is null)
+                    {
+                        return;
+                    }
+
+                    var immutableArrayTypes = systemCollectionsImmutableNamespace.GetMembers(nameof(ImmutableArray)).OfType<INamedTypeSymbol>().Where(type => type.MetadataName == typeof(ImmutableArray<>).Name).ToArray();
+                    var localSymbol = immutableArrayTypes.FirstOrDefault(type => type.ContainingAssembly.Equals(compilation.Assembly));
+                    var publicSymbol = immutableArrayTypes.FirstOrDefault(type => type.DeclaredAccessibility == Accessibility.Public);
+                    var fallbackSymbol = immutableArrayTypes.FirstOrDefault();
+                    immutableArraySymbol = localSymbol ?? publicSymbol ?? fallbackSymbol;
+                }
+
+                if (immutableArraySymbol is null)
                 {
                     return;
                 }
 
+                var immutableCollectionsAssembly = immutableArraySymbol.ContainingAssembly;
```

Comment:
📝 Changes in this analyzer were necessary to account for cases where we are compiling against implementation assemblies instead of reference assemblies.

Reply:
https://github.com/dotnet/roslyn-sdk/issues/305